### PR TITLE
Order messages thread link was modified to use symfony controller route

### LINKED ISSUE #10556: Migrate "Customer Service > Order Messages > Add new / edit order message" page
Part of Symfony migration project
Type: page

## Status

Todo:
- [x] Migrate form https://github.com/PrestaShop/PrestaShop/pull/14970

#### Controller / routing
- [ ] All controller actions are migrated and protected with `@AdminSecurity` annotations when eligible and `@DemoRestricted` annotations when eligible
- [ ] All links targeting this page are updated: either using _legacy_link or updating the code
- [ ] The new SF routes have `_legacy_link` and `_legacy_parameters` provided and explained
- [ ] Controller catches Domain Exception and translate them nicely
- [ ] Add/Edit controllers must catch global Exception in order to handle ModuleErrorException
#### Template
- [ ] Twig templates must contain relevant blocks to allow partial extension by modules
- [ ] Twig forms must have `form_rest()` and `form_errors()`
- [ ] JS being used on the page can be used as enabled extensions and use a PageMap for selectors
- [ ] Use `@PrestaShop` instead of `PrestaShopBundle` for template root path aliasing
##### Conventions
- [ ] The new controller actions comply with our naming convention
- [ ] The new SF routes names and paths comply with our naming convention
- [ ] The new Twig template names comply actions with our naming convention
- [ ] Created files (controller, JS assets, CSS assets and twig views) must be in the right folder following the convention
- [ ] CSS being added must follow theme conventions

#### Core layer
- [ ] Commands and Queries only use native types but internally use VOs
- [ ] All Handlers must have an Interface
##### Conventions
- [ ] The created SF service names comply with our naming convention
- [ ] The created files must follow the folder convention (Domain, Core, Adapter...)

#### Polishing
- [ ] The legacy controller and templates files are gone
- [ ] If PR includes a new concept/mechanism, documentation must be written
- [ ] JS and CSS assets must be compiled and up-to-date
- [ ] License headers must be up-to-date
- [ ] New hooks have been added to the hooks list and the SQL updates
- [ ] Performance must be correct: page loading time, filtering/sorting/searching response time, actions response time

#### Testing
- [ ] Commands and Queries are tested with Behat tests
- [ ] Eligible Core classes are covered with unit tests*
- [ ] Controller is covered with survival test
- [ ] (soon) Controller is covered with E2E tests using puppeeter
- [ ] If we break E2E tests because IDs change, we must update the E2E tests

*Eligible = easy to isolate and unit tests provide values (for example testing getters/setters is useless)

### PR DESCRIPTION
**Pull request by tdavidsonas88**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Order messages customer thread link was modified to use symfony controller route
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10556
| How to test?  | Rebuilding assets is required. From the Front Office go to Order history -> Details and then select product, enter message and send it. Login to the Back Office -> Notification dropdown toggle (bell icon top right corner) -> Messages -> should be able to go to the migrated controller link and see customer thread messages. If you do not see the new links, please go to Advanced Parameters -> Performance -> Clear Cache

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16678)

### CODE REVIEW

File: admin-dev/themes/new-theme/js/notifications.js [hunk 1]
```diff
@@ -63,26 +65,31 @@ const refreshNotifications = function () {
...
+        .replace(/_customer_name_/g, value.customer_name)
+        .replace(/_iso_code_/g, value.iso_code)
+        .replace(/_carrier_/g, (value.carrier !== "" ? ` - ${value.carrier}` : ""))
+        .replace(/_total_paid_/g, value.total_paid)
+        .replace(/_id_customer_/g, parseInt(value.id_customer))
+        .replace(/_company_/g, (value.company !== "" ? ` (${value.company}) ` : ""))
+        .replace(/_date_add_/g, value.date_add)
+        .replace(/_status_/g, value.status)
+        .replace(/order_url/g, `${baseAdminDir}index.php?tab=AdminOrders&token=${token_admin_orders}&vieworder&id_order=${value.id_order}`)
+        .replace(/customer_url/g, `${baseAdminDir}index.php?tab=AdminCustomers&token=${token_admin_customers}&viewcustomer&id_customer=${value.id_customer}`);
```
Comment (matks) [inline admin-dev/themes/new-theme/js/notifications.js]: Nice catch @tdavidsonas88 ! I see there is also a link using legacy customer route although Customers page has been migrated as well. Could you please do the same fix for line 85 ? 😉

Comment (tdavidsonas88) [reply admin-dev/themes/new-theme/js/notifications.js]: Thank you @matks !  committed the change for legacy customer route

File: admin-dev/themes/new-theme/js/notifications.js [hunk 2]
```diff
@@ -22,6 +22,8 @@
  * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
  * International Registered Trademark & Property of PrestaShop SA
  */
+import Router from "./components/router";
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]:
```suggestion
import Router from './components/router';
```

File: admin-dev/themes/new-theme/js/notifications.js [hunk 3]
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
...
+        .replace(/_iso_code_/g, value.iso_code)
+        .replace(/_carrier_/g, (value.carrier !== "" ? ` - ${value.carrier}` : ""))
+        .replace(/_total_paid_/g, value.total_paid)
+        .replace(/_id_customer_/g, parseInt(value.id_customer))
+        .replace(/_company_/g, (value.company !== "" ? ` (${value.company}) ` : ""))
+        .replace(/_date_add_/g, value.date_add)
+        .replace(/_status_/g, value.status)
+        .replace(/order_url/g, `${baseAdminDir}index.php?tab=AdminOrders&token=${token_admin_orders}&vieworder&id_order=${value.id_order}`)
+
+      let customerThreadId = parseInt(value.id_customer_thread);
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]:
```suggestion
      const customerThreadId = parseInt(value.id_customer_thread, 10);
```

File: admin-dev/themes/new-theme/js/notifications.js [hunk 4]
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
...
+      if (customerThreadId) {
+        let route = router.generate('admin_customer_threads_view', {customerThreadId});
+        tplReplaced = replaceGlobally(tplReplaced, 'message_url', route);
+      }
+
+      let customerId = parseInt(value.id_customer);
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]:
```suggestion
      const customerId = parseInt(value.id_customer, 10);
```

File: admin-dev/themes/new-theme/js/notifications.js [hunk 5]
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
...
+      if (customerId) {
+        let route = router.generate('admin_customers_view', {customerId});
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]: You should declare `route` and `tplReplaced` once

File: admin-dev/themes/new-theme/js/notifications.js [hunk 6]
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
...
-          .replace(/_date_add_/g, value.date_add)
-          .replace(/_status_/g, value.status)
-          .replace(/order_url/g, `${baseAdminDir}index.php?tab=AdminOrders&token=${token_admin_orders}&vieworder&id_order=${value.id_order}`)
-          .replace(/customer_url/g, `${baseAdminDir}index.php?tab=AdminCustomers&token=${token_admin_customers}&viewcustomer&id_customer=${value.id_customer}`)
-          .replace(/message_url/g, `${baseAdminDir}index.php?tab=AdminCustomerThreads&token=${token_admin_customer_threads}&viewcustomer_thread&id_customer_thread=${value.id_customer_thread}`)
-      );
+      let tplReplaced = tpl.replace(/_id_order_/g, parseInt(value.id_order))
+        .replace(/_customer_name_/g, value.customer_name)
+        .replace(/_iso_code_/g, value.iso_code)
+        .replace(/_carrier_/g, (value.carrier !== "" ? ` - ${value.carrier}` : ""))
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]:
```suggestion
        .replace(/_carrier_/g, (value.carrier !== '' ? ` - ${value.carrier}` : ''))
```

File: admin-dev/themes/new-theme/js/notifications.js [hunk 7]
```diff
@@ -60,35 +62,50 @@ const refreshNotifications = function () {
...
+        .replace(/_total_paid_/g, value.total_paid)
+        .replace(/_id_customer_/g, parseInt(value.id_customer))
+        .replace(/_company_/g, (value.company !== "" ? ` (${value.company}) ` : ""))
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]:
```suggestion
        .replace(/_company_/g, (value.company !== '' ? ` (${value.company}) ` : ''))
```

File: admin-dev/themes/new-theme/js/notifications.js (hunk 5 above)
Comment (tdavidsonas88) [reply admin-dev/themes/new-theme/js/notifications.js]: declared outside foreach, hope it is what you expected. If not, let me know

File: admin-dev/themes/new-theme/js/notifications.js [hunk 8]
```diff
@@ -64,36 +66,55 @@ let fillTpl = function (results, eltAppendTo, tpl) {
   eltAppendTo.children('.notification-elements').empty();
   if (results.length > 0) {
     eltAppendTo.removeClass('empty');
+
+    let router = new Router();
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/notifications.js]:
```suggestion
    const router = new Router();
```

Comment (PierreRambaud) [reply admin-dev/themes/new-theme/js/notifications.js]: Sorry last thing :)

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @tdavidsonas88 :)

With the PR installed, and the assets rebuild (BTW, don't hesitate to specify that rebuilding assets is required in the "How to test"), the link from the notifications still redirect me to Legacy page and not to migrated page.

Could you take a look on this bug ?

Thanks !

**Comment by tdavidsonas88** [comment]
> Hi @tdavidsonas88 :)
> 
> With the PR installed, and the assets rebuild (BTW, don't hesitate to specify that rebuilding assets is required in the "How to test"), the link from the notifications still redirect me to Legacy page and not to migrated page.
> 
> Could you take a look on this bug ?
> 
> Thanks !

Hi @Robin-Fischer-PS :)

After rebuilding the assets please check if you see in your browser sources: main.bundle.js in admin-dev/themes/new-theme/public . If you don't see it, then go to Advanced Parameters -> Performance -> Clear Cache. If that does not help and the code is merged then please let me know.

Also if the route is not found then run `bin/console fos:js-routing:dump --format=json --target=admin-dev/themes/new-theme/js/fos_js_routes.json`

**Comment by Robin-Fischer-PS** [comment]
Hi @tdavidsonas88 ! With the cache cleared it's now OK !
However, as soon as I go back to any Legacy page, the "main.bundle.js" source disappears and the link is wrong again (this happens when Cache is enabled, and also when cache is disabled).

I don't think this is OK, but maybe I'm missing something

**Comment by tdavidsonas88** [comment]
Hi @Robin-Fischer-PS @matks ! Would it be good, if I change default theme links to use these new controllers? (not sure about the backward compatibility) . Default theme has a duplicate file notifications.js which has these legacy links also.

**Comment by tdavidsonas88** [comment]
Hi @Robin-Fischer-PS @matks ! changed default theme legacy links also

File: js/admin/notifications.js [hunk 9]
```diff
@@ -116,7 +116,7 @@ function getPush()
 				// Add customers notifications to the list
 				html = "";
 				$.each(json.customer.results, function(property, value) {
-					html += "<a class='notif' href='" + value.customer_view_url + "'>";
+					html += "<a class='notif' href='"+baseAdminDir+"index.php/sell/customers/" + parseInt(value.id_customer, 10) +"/view?_token=" + token_admin_customers + "'>";
```
Comment (PierreRambaud) [inline js/admin/notifications.js]: it's not a good idea to have route in plain text in the javascript. Please use the router component instead

File: src/Core/Domain/Notification/QueryResult/NotificationResult.php [hunk 10]
```diff
@@ -119,7 +125,8 @@ public function __construct(
         string $isoCode,
         string $company,
         string $status,
-        string $dateAdd
+        string $dateAdd,
+        string $customerThreadViewUrl = ''
```
Comment (tdavidsonas88) [inline src/Core/Domain/Notification/QueryResult/NotificationResult.php:129]: @matks reviewed the solution with @rokaszygmantas , would be nice if somebody confirm this does not break backward compatibiliy as the new property was added in order to use _legacy_link in customer_threads.yml

Comment (rokaszygmantas) [reply src/Core/Domain/Notification/QueryResult/NotificationResult.php:129]: we intentionally left this `$customerThreadViewUrl` property optional in constructor, to not break BC, but maybe that's not a problem? @matks

**Comment by PierreRambaud** [comment]
Look like something went wrong with your rebase :thinking:

**Comment by tdavidsonas88** [comment]
> Look like something went wrong with your rebase

Yes, fixing I thought this goes to 1.7.7.x and rebased to it. Then reset hard to previous commit and rebased from develop

**Comment by tdavidsonas88** [comment]
> Look like something went wrong with your rebase

@PierreRambaud Rebased correctly now, fixed conflicts, ready for review; No longer blocking this one as a dev :slightly_smiling_face:

**Comment by NeOMakinG** [comment]
@tdavidsonas88 Hey, can you fix ESLint issues before we review ? Thanks :)

**Comment by tdavidsonas88** [comment]
No longer blocked by me. Label can be changed! :slightly_smiling_face:

File: admin-dev/themes/new-theme/js/notifications.js [hunk 11]
```diff
@@ -22,13 +22,14 @@
...
+import Router from './components/router';
+
 const refreshNotifications = function () {
   let timer = null;
-
   $.ajax({
     type: 'POST',
     headers: {'cache-control': 'no-cache'},
-    url: `${window.adminNotificationGetLink}&rand=${new Date().getTime()}`,
+    url: `${window.admin_notification_get_link}&rand=${new Date().getTime()}`,
```
Comment (NeOMakinG) [inline admin-dev/themes/new-theme/js/notifications.js:32]: Just a question, can't find why, why not keeping camelCase as it's cleaner ?

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
PR rebased (and approvals gone because of that)

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
I had to rebuild this PR to move its base from `develop` to `1.7.7.x` https://github.com/PrestaShop/PrestaShop/pull/19385/

**Pull request closed without merging**